modules/assim.sequential/R/downscale_function.R

```diff
@@ -1,88 +1,284 @@
+    maps <- list()
+    predictions <- list()
+    for (i in seq_along(cnn_output)) {
+      # Create a SpatRaster with the same properties as covariates
+      prediction_rast <- terra::rast(covariates)
+      
+      # Use terra::predict to apply the CNN model
+      maps[[i]] <- terra::predict(prediction_rast, model = cnn_output[[i]], 
+                                  fun = cnn_predict, 
+                                  scaling_params = scaling_params)
+      
+      # Generate predictions for testing data
+      predictions[[i]] <- cnn_predict(cnn_output[[i]], x_data[-sample, ], scaling_params)
+    }
+    
+    # Organize the results into a single output list
+    downscale_output <- list(
+      data = list(training = train_data, testing = test_data),
+      models = cnn_output,
+      maps = maps,
+      predictions = predictions,
+      scaling_params = scaling_params
+    )
+  } else {
+    stop("Invalid model_type. Please choose either 'rf' for Random Forest or 'cnn' for Convolutional Neural Network.")
   }
   
-  # Generate predictions (maps) for each ensemble member using the trained models
-  maps <- list(ncol(rf_output))
-  for (i in 1:length(rf_output)) {
-    maps[[i]] <- terra::predict(object = covariates,
-                                model = rf_output[[i]],na.rm = T)
+  # Rename each element of the output list with appropriate ensemble numbers
+  for (i in seq_along(carbon_data)) {
+    names(downscale_output$models)[i] <- paste0("ensemble", i)
+    names(downscale_output$maps)[i] <- paste0("ensemble", i)
+    names(downscale_output$predictions)[i] <- paste0("ensemble", i)
   }
   
-  # Organize the results into a single output list
-  downscale_output <- list(ensembles, rf_output, maps)
+  return(downscale_output)
+}
+
+##' @title Calculate Metrics for Downscaling Results
+##' @name calculate_metrics
+##' @author Sambhav Dixit
+##'
+##' @param downscale_output List. Output from the SDA_downscale function, containing data, models, maps, and predictions for each ensemble.
+##' @param carbon_pool Character. Name of the carbon pool used in the downscaling process.
+##'
+##' @details This function calculates performance metrics for the downscaling results. It computes Mean Squared Error (MSE), Mean Absolute Error (MAE), and R-squared for each ensemble. The function uses the actual values from the testing data and the predictions generated during the downscaling process.
+##'
+##' @description This function takes the output from the SDA_downscale function and computes various performance metrics for each ensemble. It provides a way to evaluate the accuracy of the downscaling results without modifying the main downscaling function.
+##'
+##' @return A list of metrics for each ensemble, where each element contains MAE , MSE ,R_squared ,actual values from testing data and predicted values for the testing data 
+
+calculate_metrics <- function(downscale_output, carbon_pool) {
```

Comment:
Update function name to be downscaling specific, as there are also PEcAn "metrics" functions in the benchmarking module.

Reply:
addressed the issue in the commit [06bf26b](https://github.com/PecanProject/pecan/pull/3321/commits/06bf26bf85e342b6b9189d8b924d5ec19ed1ddb0) by changing the name from calculate_metrics to SDA_downscale_metrics